### ISSUE TITLE
kvserver,cdc: high rangefeed count leads to scheduler overload

### ISSUE THREAD

**Issue by irfansharif** [cockroachdb/cockroach#96395]
**Describe the problem**

In recent incidents we've observed a high number of rangefeeds resulting in CPU scheduler overload, causing excessive scheduling latencies. These scheduling latencies in turn contribute to high foreground latencies. All this can happen even with CPU % being far from saturation (<25%). The underlying cause for all of this is the various O(ranges) goroutines we have on both the client and server side; after using MuxRangefeed, we have:
```
Client side:
  a. 1 per range in singleRangeFeed
 
    kvcoord           dist_sender_rangefeed.go:516             (*DistSender).singleRangeFeed.func3()
    kvcoord           dist_sender_rangefeed_canceler.go:100    (*stuckRangeFeedCanceler).do(*, *)
    kvcoord           dist_sender_rangefeed.go:515             (*DistSender).singleRangeFeed(#1050, {#184, *}, {{*, *, *}, {*, *, *}}, {*, ...}, ...)
    kvcoord           dist_sender_rangefeed.go:336             (*DistSender).partialRangeFeed(#1050, {#184, *}, *, {{*, *, *}, {*, *, *}}, ...)
    kvcoord           dist_sender_rangefeed.go:120             (*DistSender).RangeFeed.func1.1({#184, *})
    ctxgroup          ctxgroup.go:169                          Group.GoCtx.func1()
    errgroup          errgroup.go:57                           (*Group).Go.func1()

Server side:
  b. 1 per range blocked in RangeFeed

    kvserver          replica_rangefeed.go:254                 (*Replica).rangeFeedWithRangeID(*, #505, *, {#43378, *})
    kvserver          replica_rangefeed.go:150                 (*Replica).RangeFeed(...)
    kvserver          store.go:3082                            (*Store).RangeFeed(#1543, *, {#505, *})
    kvserver          stores.go:216                            (*Stores).RangeFeed(*, *, {#505, *})
    server            node.go:1227                             (*Node).RangeFeed(#505, *, {#505, *})
    rpc               context.go:753                           internalClientAdapter.RangeFeed.func1()

  c. 1 per range in the processor
  
    rangefeed         processor.go:269                         (*Processor).run(*, {#5, *}, *, *, #1177)
    rangefeed         processor.go:222                         (*Processor).Start.func1({#5, *})
    stop              stopper.go:494                           (*Stopper).RunAsyncTaskEx.func2()

  d. 1 per range in the registration output loop

    rangefeed         registry.go:310                          (*registration).outputLoop(*, {#485, *})
    rangefeed         registry.go:335                          (*registration).runOutputLoop(*, {#5, *}, *)
    rangefeed         processor.go:297                         (*Processor).run.func1({#5, *})
    stop              stopper.go:494                           (*Stopper).RunAsyncTaskEx.func2()
```

Known incidents:
- https://github.com/cockroachlabs/support/issues/2036
- https://github.com/cockroachlabs/support/issues/1997

<details>
<summary>To understand, consider these metrics from an actual incident.</summary>

It ends up looking something like:

![image](https://user-images.githubusercontent.com/10536690/216187221-8947cd2a-9eaf-4a96-a773-f43587295d5e.png)

![image](https://user-images.githubusercontent.com/10536690/216187285-a3c1c7dc-7356-4d42-8e69-308fca1c8ffc.png)

In these incidents, we also see in AC kicking in, but helplessly since it's unable to keep runnable g's stable under it's configured threshold. Consider the KV-Queue on n113 is firing during this period for example, and it observing wild swings between used and total slots (where used >> total at some points), so there's a lot of spurious goroutine wakeups. p75 AC delay is quite high (1s+), across all CPU queues. Dividing the delay rate of n113's KV queue by it's admission rate, we get a mean latency of 277.78ms per request, which is very high. 

![image](https://user-images.githubusercontent.com/10536690/216187785-6f8eb40b-950e-409d-94c4-8974bd3af5f2.png)

![image](https://user-images.githubusercontent.com/10536690/216187855-f81d7862-7ac8-433d-940e-8f11bc41798b.png)

</details>

**To Reproduce**

Run a changefeed over a table with many underlying ranges. We should start with adding a roachtest reproducing this goroutine count induced latency hit; it's tracked here: https://github.com/cockroachdb/cockroach/issues/95236.

**Describe alternatives you've considered**

- Reducing the frequency of these goroutine wakeups has helped in incidents, for ex. by increasing `kv.rangefeed.closed_timestamp_refresh_interval` to be 2s instead of 200ms (read from `kv.closed_timestamp.side_transport_interval`).
- Reducing the number of ranges over which we need to establish rangefeeds. Depending on the usecases, if we could either leverage the predicate pushdown we've introduced in https://www.cockroachlabs.com/docs/stable/cdc-transformations.html for primary keys, or add smarts to also consider secondary indexes.
- @miretskiy has a prototype where he shuts down quiescent ranges for some period of time, reducing not only client goroutines but also the server-side ones.

Jira issue: CRDB-24101

Epic CRDB-26372

**Comment by ajwerner** [comment]
One thing I thing matters is how often we wake these goroutines up. Today we wake them up once per checkpoint and that checkpoint frequency is a function of the side transport interval and the rate of writes to the range. My expectation is that the latency impacts would be less severe if we reduced the rate of state changes of these goroutines. There's a separate issue about more fine-grained control of the rate of checkpoint propagation. 

We should study how that affects latency. If we reduced the rate of wakeups from like 5Hz per per-range goroutine to like .1Hz, would things be way better? If the events are all at the same time, maybe not as much.

**Comment by ajwerner** [comment]
https://github.com/cockroachdb/cockroach/issues/67632 is the issue I was thinking of.

**Comment by irfansharif** [comment]
> My expectation is that the latency impacts would be less severe if we reduced the rate of state changes of these goroutines.

Your expectation is right. In one of these incidents above bumping `kv.rangefeed.closed_timestamp_refresh_interval` to 2s effectively dampened this latency effect. I'll note that this issue is primarily a placeholder, to point to when we start seeing these cases, not something to immediately fix by reducing g count if the workarounds (reducing checkpoint frequency, using CDC predicate pushdowns) are acceptable. I suspect they will be.

**Comment by irfansharif** [comment]
This is somewhat unrelated but similar looking symptoms. We've seen incidents where an excessive number of goroutines due to an excessive number of SQL connections cause similar scheduler overload and thrashing in AC's slot adjustment. See https://github.com/cockroachlabs/support/issues/2045#issuecomment-1412479981, and this [internal discussion](https://cockroachlabs.slack.com/archives/C01SRKWGHG8/p1675366778810919). In these incidents the user was outside far outside prescribed limits for # of SQL connections, but the induced AC delay seemed initially suspect. Copying over a snippet:

```
This workload is using what feels like an unreasonable amount of concurrency/active SQL connections. The slot-based control mechanism wasn’t built with this operating point in mind. So it’s thrashing. But even if AC was switched off, this high concurrency is going to cause lots of scheduling delay. So your end latencies are still going to be bad. AC is just a bit ineffective here — it’s trying to shift all scheduling delay into its own queues but that delay is very erratic because of this very high concurrency, so it’s making a bad situation possibly slightly worse. But it’s still bad to begin with. 
```

**Comment by sumeerbhola** [comment]
- Is the rangefeed work that is done when the checkpoint is published only use CPU or does it also need to do IO?
- If it is all CPU, it would be reasonable to say that we should size a goroutine pool that is say equal to number of processors, so that if rangefeed processing needs to use 100% of the CPU, it can (this is not elastic work). If it is not all CPU, but we can safely say that the IO/CPU time ratio is unlikely to be > K, then we can size the pool to be (K+1)*numProcs. This will add at most K+1 to the runnable goroutine count, which is better than the current behavior.
- It is important that the normal behavior is for this work to be smeared across time instead of being bursty. Is there a `rangefeed.Processor` per range (I couldn't figure it out with a cursory examination of the code, since each `registration` object has a span)? If the processor is per range, is there something that is causing all the different range processors to `publishCheckpoint` at the same time?

**Comment by ajwerner** [comment]
> when the checkpoint is published only use CPU or does it also need to do IO?

The IO that happens is sending on a gRPC stream.

>  If the processor is per range, is there something that is causing all the different range processors to publishCheckpoint at the same time?

It is per range. It is bursty. We have a thing which deals with progressing the closed (and thus resolved) timestamp for quiescent ranges. It has a fixed tick interval. When it publishes, it publishes to all ranges. The sad part is that we want to deliver these messages in a timely manner, so smearing over time is difficult.

I think the "right" answer here is to leverage the `MuxRangeFeed` abstraction and find a way to not turn each tick into a gRPC message and wake-up per range*clients but rather per client. This might fall out naturally if we try to re-arrange the threading model for these ranges. Right now, the way the code is written, it's difficult to talk about how to do that.


Today (assuming MuxRangeFeed, which is sort of required) it's like ([uml](www.plantuml.com/plantuml/png/NOxFQlCm4CJlVeh5f--314AFFPH0eRT2QN8hfGXh90lqn-okqlJj6zauXbwaGJFwpHn53MjDiRi91VNlnYczSHaW4hBloGZntNsEHM3f7194JPhe52pep2RBL5Zdj6Z7yP5AfcyE2jcHBrogXWnoN1Bf26BZz3ICi3M2QIgbLRHF0mWNy0_rqj2weYba1LzCxEcVXJCrAZqS3ad-RykpnmjWr_c5b_LkekvsD3fe0XEtPCUHzlkdPi-IWn_rOmpM8d_UtaKxqItXfaN3t_lWWoWR3IMtejbZ-ArdY05PjyhcpfHKhr_iLj1hlTwtDz1gBzTVt8vaD0nBPk0Yif2bEoBRck8l)):
![NOxFQlCm4CJlVeh5f--314AFFPH0eRT2QN8hfGXh90lqn-okqlJj6zauXbwaGJFwpHn53MjDiRi91VNlnYczSHaW4hBloGZntNsEHM3f7194JPhe52pep2RBL5Zdj6Z7yP5AfcyE2jcHBrogXWnoN1Bf26BZz3ICi3M2QIgbLRHF0mWNy0_rqj2weYba1LzCxEcVXJCrAZqS3ad-RykpnmjWr_c5b_LkekvsD3fe0XEtPCUH](https://user-images.githubusercontent.com/1839234/216657908-6ede68cc-ae12-49f4-aec7-aad543136e78.svg)

**Comment by erikgrinaker** [comment]
We've been hitting similar O(range) performance problems with expiration lease extensions and Raft processing too. Instead of building increasingly sophisticated machinery to quiesce and coalesce work across ranges, which comes with significant complexity and availability tradeoffs, we're strongly considering to instead address this by bumping the range size to e.g. 4 GB and keeping individual nodes below 1000 ranges or so, bounding the cost of per-range work. This would also significantly help with the rangefeed issues here, at least for the server-side work (we'd still have to avoid having one client goroutine for each rangefeed across the entire cluster).

**Comment by sumeerbhola** [comment]
I wouldn't necessarily conflate this burstiness issue with total size of work issues in the context of "O(range) performance problems with expiration lease extensions and Raft processing too".

**Comment by erikgrinaker** [comment]
Fair, even 1000 goroutines would likely cause scheduling issues due to the checkpoints. @ajwerner says on Slack:

>we’re doing a fan-out just to fan back in [...] My sense is that the right hand on the wheel could refactor the server side to not be O(ranges) goroutines but rather a pool of goroutines like the raft scheduler for the `rangefeed.Processor` and then pushing all of the per-client work onto the goroutine due to that client.

I'm inclined to agree with this. We should aim both to have a rangefeed scheduler using a bounded goroutine pool, and to coalesce the work needed to emit checkpoints (in particular the noop case where most ranges haven't seen any changes).

**Comment by erikgrinaker** [comment]
This has mostly been addressed by the rangefeed scheduler in #107553 and the mux rangefeed protocol. The one remaining wrinkle is the rangefeed registrations, which we didn't get in for 23.2, but will address for 24.1 -- I've split that out to #110432.

The scheduler and 3-second closed timestamp interval gets us most of the way there, the registrations mostly affect tail latency. Here are some KV workload benchmarks with a concurrent changefeed using various rangefeed configurations:

```
kv0 read-only, 100k ranges, 5 nodes, 3s closed timestamps

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1200.0s        0       19194765        15995.6     20.0      1.9     33.6    352.3   9663.7  default
 1200.0s        0       44620243        37183.5      8.6      1.4     19.9    142.6   2415.9  closedts=3s
 1200.0s        0       53535090        44612.5      7.2      1.4     22.0    109.1   1677.7  mux+closedts=3s
 1200.0s        0       79908559        66590.4      4.8      1.3     24.1     37.7   1140.9  scheduler+mux+closedts=3s
 1200.0s        0       78458847        65382.3      4.9      1.3     24.1     33.6    486.5  scheduler+mux+closedts=3s+registrations
 1200.0s        0       82444569        68703.7      4.7      1.3     22.0     31.5    604.0  no changefeed

kv0 write-only, 100k ranges, 5 nodes

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1200.0s      258        5998370         4998.6     64.0     11.0    151.0   1543.5  12884.9  default
 1200.0s        0       14504448        12087.0     26.5      7.3     79.7    436.2   6174.0  closedts=3s
 1200.0s        0       25664364        21386.9     15.0      6.6     60.8    113.2   3355.4  mux+scheduler+closedts=3s
 1200.0s        0       25896305        21580.2     14.8      7.1     56.6    100.7   2952.8  mux+scheduler+closedts=3s+registrations
 1200.0s        0       29986637        24988.8     12.8      6.3     48.2     88.1   2281.7  no changefeed
```
